documentation/ArrayExpressionLoweringToFIR.md

```diff
@@ -0,0 +1,315 @@
+to achieve short-circuiting.
+
+These rules should guarantee that we do not have to track the CFG state while lowering an expression other than the current insertion point
+(and maybe some places to load symbols. allocate temps...).
+
+### An array shape does not imply an expression is elemental
+All expressions with an array shape are not elemental, for instance `foo(array)` where `foo` is a user
+function that is not elemental cannot be replaced by call to `foo` on each `array` element.
+
+### Some non-elemental expression can be seen as an elemental expression and an affine map
+As stated in https://github.com/flang-compiler/f18/blob/master/documentation/ArrayComposition.md,
+transformational functions are not elemental but they many of them can simply be viewed as be viewed as
+changing the way their array arguments are to be accessed (i.e/ they are functions operating on the indexes and can be seen in MLIR as affine maps).
+
+### Do not early optimize but do not prevent later optimizations
+Expression lowering from the parse-tree to FIR is not meant to be an optimization pass,
+it should only produce FIR that can be optimized later. For instance, if as stated in ArrayComposition.md,
+it is an end goal to avoid unnecessary temporary arrays, lowering is not the place to perform deep temporary removal optimization.
+It is not entirely clear yet what "can be optimized later" implies.
+The strategy should be to start lowering in a simple way and if it turns out that some optimizations are later not possible, it should be
+preferred to find a way to convey the information that is missing in FIR to enable the optimization rather than to perform the optimization
+while lowering.
+
+### Not all elemental operation should be decomposed
+`merge(x, y, b)` when b is scalar does not require a loop to be implemented.
+It can be implemented as single branch to load the right value or even better, a merge
+on the memory reference.
+
+Question: If lowering were to emit a loop, would later optimization be able to transform this to a simple branch ?
+
+Note: This anyway falls in the short-circuiting cases.
+
+Also, the simple expression `array` should not be broken into a loop if this is not needed,
+it has the value `fir.load %array`.
+
+## Expression Lowering Design
+
+### Proposed Generated FIR structure for one array expression
+1. Load all required symbols (+)
+2. Compute shape (if not compiled time constant)
+3. Short-circuit if null shape
+4. Allocate tmp result (stack or heap?) (+)
+5. Compute scalar operands (+)
+6. Short-circuit if scalar operands are enough for result
+7. Compute array sub-expressions (+)
+8. Loop on element (+)
+9. short-circuit targets 
+
+Steps 2. to 9. are repeated for each "non-inlined" array sub expression.
+By "non-inlined", it is meant that it is a sub-expression that is computed in its own loop and
+is associated its own memory storage.
```

Comment:
Yes, that is the naive expression lowering proposal that does not try to group different elemental sub-expression into a single `fir.loop`. It is indeed not my favorite. That one assumes FIR transformations will later take care of merging all the iteration spaces.
I am adding illustrations so that it is clearer.

Reply:
I think you already know this, but for completeness, I'll write it up anyway.

We need to keep in mind the distinction between structure-preserving composition vs. function-preserving composition.  For a trivial example, let's say we have an expression
```
   a = b + c * d
```
If this is rank 0, we'd almost surely lower this to a representation that preserves the _functional_ semantics of the expression.
```
  %c = load %c.addr
  %d = load %d.addr
  %1 = mul %c, %d
  %b = load %b.addr
  %2 = add %b, %1
  store %2 to %a.addr
```
If we instead wanted to preserve the _structure_ of the intermediate values, we'd do something like
```
  %c = load %c.addr
  %d = load %d.addr
  %1 = mul %c, %d
  %tmp.addr = alloca T
  store %1 to %tmp.addr
  %b = load %b.addr
  %tmp = load %tmp.addr
  %2 = add %b, %tmp
  %tmp2.addr = alloca T
  store %2 to %tmp2.addr
  %tmp2 = load %tmp2.addr
  store %tmp2 to %a.addr
```
If we extend this example to an expression with a rank > 0, all of these operations become "little loops".  But we don't have to preserve the _structure_ (call it a three-address loop code), per se. Temporaries where the effects are not observable can be merged into a (slightly) bigger loop.
```
  %tmp.addr = alloca T
loop1:
  %c = load %c.addr[%i]
  %d = load %d.addr[%i]
  %1 = mul %c, %d
  store %1 to %tmp.addr[%i]
end-loop1:
  %tmp2.addr = alloca T
loop2:
  %b = load %b.addr[%i]
  %tmp = load %tmp.addr[%i]
  %2 = add %b, %tmp
  store %2 to %tmp2.addr[%i]
end-loop2:
loop3:
  %tmp2 = load %tmp2.addr[%i]
  store %tmp2 to %a.addr[%i]
end-loop3:
```
can more simply be
```
loop:
  %c = load %c.addr[%i]
  %d = load %d.addr[%i]
  %1 = mul %c, %d
  %b = load %b.addr[%i]
  %2 = add %b, %1
  store %2 to %a.addr[%i]
end-loop:
```

Reply:
Thanks for taking the time to detail this !  It puts some better words and examples on the alternatives that were presented in the document. The new document posted on flang tries to use this function vs structure distinction.